TRUNK-5473 - Refactor PersonValidator#validate method

fruether opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'TRUNK-5111 Replace use of deprecated isVoided' -->
TRUNK-5473 - Refactor PersonValidator#validate method
## Description of what I changed
This PR is highly inspired by (https://github.com/openmrs/openmrs-core/pull/2804) Thanks for the good previous work.
- PersonValidator#validate  simplified by creating sub-methods for validatePersonAddresses, validatePersonNames and validatePersonHasAtLeastOneNonVoidedName
- I added all validate methods into PersonValidator and not as suggested by previous work into the Person class
- I am not sure how the PersonValidator should deal with a Null in a personAddress.  I made it to simply continue without failing. See here fore more details: https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/validator/PersonAddressValidator.java#L67


## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/TRUNK-5473
## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [X] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [X] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

dkayiwa commented:
The link to the ticket does not look correct.

fruether commented:
@dkayiwa I changed the link. It works now

fruether commented:
@dkayiwa, @chankruze  any updates regarding this PR :) ?

dkayiwa commented on api/src/main/java/org/openmrs/validator/PersonValidator.java:
```diff
@@ -68,41 +68,79 @@ public void validate(Object target, Errors errors) {
 		}
 		
 		Person person = (Person) target;
+		if (person.getVoided()) {
```
Is this refactoring or new functionality?

fruether replied:
No new functionality. I just rearrange this line of code. It was previously in the end. Now I moved it up to make the code easier/more readable:

See line 103: https://github.com/openmrs/openmrs-core/pull/2953/commits/69fba31b11fd3454630bb643e9dcb1685833a079#diff-7014d6c9eb6084f32ffc5c659e8ee524L103

dkayiwa replied:
We usually do such formatting changes in separate pull requests.

fruether replied:
I did so because for me it seemed to be a valid refactoring operation. Do you want this Pull Requests to be split up?

dkayiwa replied:
Because any refactoring has potential to introduce unintended bugs, i find it easier and cleaner to focus on only what the ticket was meant for. You can always create more tickets for further improvements.

fruether replied:
My understanding was that refactoring to make a method simpler would also mean to reduce over-complex if conditions and make the code easier by layout. Anyways I removed the difference in the if condition and change in order. It should now only contain the new methods.

dkayiwa commented:
@fruether these changes seem to be dealing with more than what the ticket description requires.

fruether commented:
@dkayiwa the issue requested a refactoring. I just did a refactoring. As part of the refactoring I moved code up/down and eventually created new methods to make everything more simple. Therefore,  did not change any behaviour. The code and functionality is the same. Just hopefully easier to read.

Can you explain more in depth why you think I did more?


Furthermore, I noticed a strange behaviour which I documented in the issue bur did not change at all: https://issues.openmrs.org/browse/TRUNK-5473?focusedCommentId=256840&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-256840

fruether commented on api/src/test/java/org/openmrs/validator/PersonValidatorTest.java:
```diff
@@ -231,5 +234,25 @@ public void validate_shouldNotFailWhenDeathCauseNotCodedIsSet() {
 	}
+	
+	@Test
+	public void validate_shouldNotFailWhenOneAddressIsNull() {
```
@dkayiwa should that test be removed as well?  Especially in context of this content I think it should be part of this PR: https://issues.openmrs.org/browse/TRUNK-5473?focusedCommentId=256840&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-256840

dkayiwa replied:
Tests to catch possible refactoring bugs, can stay. 😊

dkayiwa commented on api/src/main/java/org/openmrs/validator/PersonValidator.java:
```diff
@@ -111,8 +151,6 @@ else if(person.getCauseOfDeath() == null && person.getCauseOfDeathNonCoded() ==
 				errors.rejectValue("causeOfDeath", "Person.dead.causeOfDeathAndCauseOfDeathNonCodedNull");
 			}
 		}
-		
```
Did you intentionally remove this?

fruether replied:
I did not remove line 114 and 115. I put it down where it was before I changed the code according to the previous pull request. Git Change Log does not recognize it this way though:
https://github.com/openmrs/openmrs-core/pull/2953/commits/4c40dade91b459e578231a2e593828832a5b51d9#diff-7014d6c9eb6084f32ffc5c659e8ee524R87

dkayiwa commented on api/src/main/java/org/openmrs/validator/PersonValidator.java:
```diff
@@ -69,40 +69,80 @@ public void validate(Object target, Errors errors) {
-				ValidationUtils.invokeValidator(personAddressValidator, address, errors);
+				personAddressValidator.validate(address, errors);
+			}
+			catch (NullPointerException exception) {
```
Are you intentionally ignoring the null pointer exception?

fruether replied:
Yes. Reason (1) To make explicit that a exception is thrown if the value is null. That fact is pretty much hidden by the finally block. I think that overcomplicates the code understanding  (2) To not change the behaviour of the code. I also added a test case to prove:

The issue lies in the fact that personAddressValidator validator does check if the value is null and then throws an exception. That is not how the other classes behave: https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/validator/PersonAddressValidator.java#L67

dkayiwa replied:
I would avoid changing behavior in a refactoring pull request. Even with the test, you never know what bugs may be introduced which the test does not catch.

fruether replied:
If I remove the catch() I also have to remove the test case I added. Since it is some weeks ago I forgot that I added this catch() since the newly added test case of me would not work otherwise. Which means I changed the behaviour. I assume you want the test case & try/catch removed?

dkayiwa replied:
Whatever it takes to restore the original behaviour.

fruether commented on api/src/test/java/org/openmrs/validator/PersonValidatorTest.java:
```diff
@@ -231,5 +234,28 @@ public void validate_shouldNotFailWhenDeathCauseNotCodedIsSet() {
+		try {
+			personValidator.validate(person, errors);
+		}catch(NullPointerException np) {
+			Assert.assertNotNull(np);
```
@dkayiwa I reverted the old state. I would suggest to create a new ticket to fix this behaviour if you agree that this is not intended.

dkayiwa replied:
Feel free to create a the new ticket.

fruether replied:
@dkayiwa are there any more issues now stopping the merging? As far as I am know all your comments are resolved now :)

fruether commented:
@dkayiwa All your comments should be applied. Anything more open regarding this issue?

dkayiwa commented:
@fruether did you check if sonar is now happy with this method?

fruether commented:
I do not have the Developer Edition of Sonar. That's why I can't analyze
this branch.
However, the new introduced functions move complex parts out of the method
and hence Cyclomatic Complexity is at least reduced by 3 points so it
should be <= 10.
It could have been reduced a bit more by the changes that were declined
during the review process since less conditions were in the if clause .

Am Sa., 20. Juli 2019 um 22:45 Uhr schrieb dkayiwa <notifications@github.com
>:

> @fruether <https://github.com/fruether> did you check if sonar is now
> happy with this method?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-core/pull/2953?email_source=notifications&email_token=AAMEOACR6PIPV6565R44BN3QAN2PXA5CNFSM4HR5CTB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2NVWQQ#issuecomment-513497922>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMEOAA42M4DFYFJKKQTSFDQAN2PXANCNFSM4HR5CTBQ>
> .
>

dkayiwa commented:
@fruether did you get a chance to look at this? https://wiki.openmrs.org/display/docs/Sonar

fruether commented:
@dkayiwa I am not allowed to execute sonar:

> ERROR] Failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.4.0.905:sonar (default-cli) on project openmrs: You're only authorized to execute a local (preview) SonarQube analysis without pushing the results to the SonarQube server. Please contact your SonarQube administrator. -> [Help 1]

dkayiwa commented on api/src/test/java/org/openmrs/validator/PersonValidatorTest.java:
```diff
@@ -231,5 +234,28 @@ public void validate_shouldNotFailWhenDeathCauseNotCodedIsSet() {
+	public void validate_shouldNotFailWhenOneAddressIsNull() {
+		Person person = new Patient(1);
+		Set<PersonAddress> addresses =  new HashSet<>();
+		addresses.add(null);
```
What use case would lead to a null address added to the collection?

dkayiwa commented on api/src/test/java/org/openmrs/validator/PersonValidatorTest.java:
```diff
@@ -231,5 +234,28 @@ public void validate_shouldNotFailWhenDeathCauseNotCodedIsSet() {
+		
+		try {
+			personValidator.validate(person, errors);
+		}catch(NullPointerException np) {
```
Are you trying to test that a null pointer exception should be thrown when there is a null in the address collection?

fruether commented:
I created this test case to illustrate that in case of a null address added
the program terminates normally.

This is to show that the PersonAdress validation deals with null values in
this way.

dkayiwa <notifications@github.com> schrieb am So., 21. Juli 2019, 20:05:

> *@dkayiwa* commented on this pull request.
> ------------------------------
>
> In api/src/test/java/org/openmrs/validator/PersonValidatorTest.java
> <https://github.com/openmrs/openmrs-core/pull/2953#discussion_r305629451>:
>
> > @@ -231,5 +234,28 @@ public void validate_shouldNotFailWhenDeathCauseNotCodedIsSet() {
>
>  		Assert.assertFalse(errors.hasErrors());
>  	}
> +	
> +	@Test()
> +	public void validate_shouldNotFailWhenOneAddressIsNull() {
> +		Person person = new Patient(1);
> +		Set<PersonAddress> addresses =  new HashSet<>();
> +		addresses.add(null);
>
> What use case would lead to a null address added to the collection?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-core/pull/2953?email_source=notifications&email_token=AAMEOAGZEHZ46YYC27X2GMLQASQNJA5CNFSM4HR5CTB2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB7CK5TQ#pullrequestreview-264548046>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMEOAHNXFIELVEEHEW2OC3QASQNJANCNFSM4HR5CTBQ>
> .
>

fruether commented:
Yes. That is basically related to the ticket I created yesterday.

It shows that the validation throws an exception opposed to the other
attributes that log the error instead of throwing the exception.

I basically use this test to document a behaviour that seemed to be strange
to me and at the same time proof that I did not change the program semantics

dkayiwa <notifications@github.com> schrieb am So., 21. Juli 2019, 20:06:

> *@dkayiwa* commented on this pull request.
> ------------------------------
>
> In api/src/test/java/org/openmrs/validator/PersonValidatorTest.java
> <https://github.com/openmrs/openmrs-core/pull/2953#discussion_r305629494>:
>
> > +		Set<PersonAddress> addresses =  new HashSet<>();
> +		addresses.add(null);
> +		
> +		person.setDead(true);
> +		person.setCauseOfDeathNonCoded("Some text describing Cause of Death");
> +		person.setPersonVoided(true);
> +		person.setPersonVoidReason("voidReason");
> +		person.setGender("g");
> +		person.setAddresses(addresses);
> +		
> +		Errors errors = new BindException(person, "patient");
> +		PersonValidator personValidator = new PersonValidator();
> +		
> +		try {
> +			personValidator.validate(person, errors);
> +		}catch(NullPointerException np) {
>
> Are you trying to test that a null pointer exception should be thrown when
> there is a null in the address collection?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-core/pull/2953?email_source=notifications&email_token=AAMEOAEHE3LWHLGKTVOMBXLQASQRPA5CNFSM4HR5CTB2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB7CK57I#pullrequestreview-264548093>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMEOAFNWNFL3TLF3TK5RSTQASQRPANCNFSM4HR5CTBQ>
> .
>

dkayiwa commented:
I do not see value in a test for a null address passed in the addresses collection. I would expect tests for real expected functionality.

fruether commented:
Its not for me to decide what behaviour is intended and what is not. So I
just created the test case to document the current behaviour as of proof
that refactoring did not change anything.

 For newly created ticket this test case could be used as a false positive.

I will remove it later and re-commit

dkayiwa <notifications@github.com> schrieb am So., 21. Juli 2019, 20:26:

> I do not see value in a test for a null address passed in the addresses
> collection. I would expect tests for real expected functionality.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-core/pull/2953?email_source=notifications&email_token=AAMEOAHNVWLCRDH66GCNVTLQASS57A5CNFSM4HR5CTB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2OI5JQ#issuecomment-513576614>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMEOAGNAIRPU47IURZMQ3TQASS57ANCNFSM4HR5CTBQ>
> .
>

dkayiwa commented:
Is your asserting in an exception handler to confirm that a null pointer exception is thrown when null is passed in the address collection?

fruether commented:
Let me try to be a bit more clear:
_**Test Case:**_  This test case validates that in case a PersonAddress is null a NullPointerException is thrown. This is the behaviour of the code before and after my refactoring.

_**Normal behaviour of error cases**_: For me it seems to be odd that a NullPointerException is thrown since it is neither part of the validate method description nor of the method definition. In fact, all other validations errors are dealt with by  errors.hasFieldErrors

_**History of my refactoring**_: Since I found that behaviour very odd and un-expecting I added during the refactoring process a try_catch_ block to the method to catch the NullPointerException and log that it was failing. [Where I would expect a  catch](https://github.com/openmrs/openmrs-core/pull/2953/files/627fe668ebfd9259df0b4d148df5155c574fc440#diff-7014d6c9eb6084f32ffc5c659e8ee524R129)
However, we agreed that we should not change the methods behaviour in refactoring. That is why I removed the catch block but sticked with the "Invered" test case (it is successful if the exception is thrown) to document the behaviour and proof that refactoring did not change the business logic

dkayiwa commented:
If your test is about expecting a null pointer exception, did you evaluate using this? @Test(expected = NullPointerException.class) or @Rule ExpectedException

fruether commented:
@dkayiwa I used now: @test(expected = NullPointerException.class)

dkayiwa commented:
Was this test passing before your refactoring changes?

fruether commented:
Good that you asked. Before my refactoring it was: 				ValidationUtils.invokeValidator(personAddressValidator, address, errors);

Isn't it odd that sometimes ValidationUtils is used and sometimes it is not?


That is why IllegalArgumentException was thrown not NullPointerException. However, good that I added the test to log the behaviour

dkayiwa commented:
The fact that your test did not fail before your refactoring changes means you are not testing the original behavior

fruether commented:
Yes. That is unfortunately the case. It was an error that is fixed now.

Am So., 21. Juli 2019 um 23:38 Uhr schrieb dkayiwa <notifications@github.com
>:

> The fact that your test did not fail before your refactoring changes means
> you are not testing the original behavior
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-core/pull/2953?email_source=notifications&email_token=AAMEOACPT5HRZL37SDA6JDLQATJLZA5CNFSM4HR5CTB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2OMDTY#issuecomment-513589711>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMEOAFFP23GIRE6KH536ZLQATJLZANCNFSM4HR5CTBQ>
> .
>

dkayiwa commented:
It gets even more confusing when your test name says should not fail but then expect an exception.

fruether commented:
The test that I added now is neither failing before nor after the
refactoring. The exception was added to the correct type as well as the
method call of validate was re-stored.

Am So., 21. Juli 2019 um 23:40 Uhr schrieb Frederik Rue <
freddy92.online@gmail.com>:

> Yes. That is unfortunately the case. It was an error that is fixed now.
>
> Am So., 21. Juli 2019 um 23:38 Uhr schrieb dkayiwa <
> notifications@github.com>:
>
>> The fact that your test did not fail before your refactoring changes
>> means you are not testing the original behavior
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/openmrs/openmrs-core/pull/2953?email_source=notifications&email_token=AAMEOACPT5HRZL37SDA6JDLQATJLZA5CNFSM4HR5CTB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2OMDTY#issuecomment-513589711>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAMEOAFFP23GIRE6KH536ZLQATJLZANCNFSM4HR5CTBQ>
>> .
>>
>

dkayiwa commented:
That does not resolve the confusion which comes from a test whose name says should not fail but at the same time expect an exception.

fruether commented:
I renamed it just now to: validate_shouldThrowExceptionWhenAddressIsNull

dkayiwa commented on api/src/main/java/org/openmrs/validator/PersonValidator.java:
```diff
@@ -69,23 +69,58 @@ public void validate(Object target, Errors errors) {
+		validatePersonNames(person, errors);
+		validatePersonAddresses(person, errors);
+		
+		if (!person.getVoided() && !validatePersonHasAtLeastOneNonVoidedName(person)) {
```
Did you intentionally change the order of the validation of names and addresses?

fruether replied:
In Line 74 personName is validated. in Line 89 personAdress in the original code. I am not sure what you mean with change of order?

dkayiwa replied:
You are validation person addresses on line 73 before the names from line 75

fruether replied:
Reason was that for me the code is more readable if both method calls followed each other.

dkayiwa replied:
Changing the order may result into unintended side effects from the caller. So i would be cautious not to do that in this pull request. Am sure you have ever seen bug or unexpected behavior introduced by a mere change of the order for a few lines of code. So let us be as surgical as possible. 😊

fruether replied:
I reverted the order
